File: epi_judge_java/epi/PrimitiveMultiply.java
package epi;

import epi.test_framework.EpiTest;
import epi.test_framework.GenericTest;

public class PrimitiveMultiply {

    @EpiTest(testDataFile = "primitive_multiply.tsv")
    public static long multiply(long x, long y) {
        long result = 0;

        while (x != 0) {
            if ((x & 1) != 0) {
                result = add(result, y);
            }

            x >>= 1;
            y <<= 1;
        }

        return result;
    }

    private static long add(long x, long y) {
        long result = 0;
        long carry = 0;
        long xCopy = x;
        long yCopy = y;
        long pow = 1;

        while (xCopy != 0 || yCopy != 0) {
            long xBit = x & pow;
            long yBit = y & pow;
            long tempCarry = (xBit & yBit) | (xBit & carry) | (yBit & carry);
            result |= xBit ^ yBit ^ carry;
            carry = tempCarry << 1;

            pow <<= 1;
            xCopy >>= 1;
            yCopy >>= 1;
        }

        return result | carry;
    }

    public static void main(String[] args) {
        System.exit(
                GenericTest
                        .runFromAnnotations(args, "PrimitiveMultiply.java",
                                new Object() {
                                }.getClass().getEnclosingClass())
                        .ordinal());
    }
}

File: epi_judge_cpp/primitive_multiply.cc
#include "test_framework/generic_test.h"

unsigned long long int Add(unsigned long long int x, unsigned long long int y);

unsigned long long Multiply(unsigned long long x, unsigned long long y) {
    unsigned long long result = 0;

    while (x) {
        if (x & 1) {
            result = Add(result, y);
        }

        x >>= 1;
        y <<= 1;
    }

    return result;
}

unsigned long long int Add(unsigned long long int x, unsigned long long int y) {
    unsigned long long result = 0;
    unsigned long long carry = 0;
    unsigned long long k = 1;

    int shift = 0;

    while ((x >> shift) || (y >> shift)) {
        auto xk = x & k;
        auto yk = y & k;

        auto tempCarry = (xk & yk) | (xk & carry) | (yk & carry);
        result |= xk ^ yk ^ carry;

        carry = tempCarry << 1;
        k <<= 1;
        shift++;
    }

    return result | carry;
}

int main(int argc, char *argv[]) {
    std::vector<std::string> args{argv + 1, argv + argc};
    std::vector<std::string> param_names{"x", "y"};
    return GenericTestMain(args, "primitive_multiply.cc",
                           "primitive_multiply.tsv", &Multiply,
                           DefaultComparator{}, param_names);
}
